### PR TITLE
Track "active" WhatsApp users and implement blocking when reaching the limit

### PR DESCRIPTION
**Pull request by Half-Shot**
This PR intends to do two things:
- Store and track activity metrics for whatsapp users, using a metric of minimum days before they are considered active.
- Block the bridge from sending/receiving traffic when the cap has been reached for active users.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -270,7 +276,39 @@ func (bridge *Bridge) Init() {
 	bridge.MatrixHandler = NewMatrixHandler(bridge)
 	bridge.Formatter = NewFormatter(bridge)
 	bridge.Crypto = NewCryptoHelper(bridge)
-	bridge.Metrics = NewMetricsHandler(bridge.Config.Metrics.Listen, bridge.Log.Sub("Metrics"), bridge.DB)
+	bridge.Metrics = NewMetricsHandler(bridge.Config.Metrics.Listen, bridge.Log.Sub("Metrics"), bridge.DB, bridge.PuppetActivity)
+}
+
+func (mh *Bridge) UpdateActivePuppetCount() {
+	if mh.Config.AppService.Limits.MinPuppetActiveDays == 0 {
```
Comment (jaller94) [inline main.go:283]: I'm a bit confused as to why 0 is not a valid value?
`1` day seems to only count them after 24 hours, but what if I want users to be counted immediately?
Currently, it shuts tracking off.

File: metrics.go [hunk 2]
```diff
@@ -103,6 +106,14 @@ func NewMetricsHandler(address string, log log.Logger, db *database.Database) *M
 			Name: "whatsapp_puppets_total",
 			Help: "Number of WhatsApp users bridged into Matrix",
 		}),
+		activePuppetCount: promauto.NewGauge(prometheus.GaugeOpts{
+			Name: "whatsapp_active_puppets_total",
+			Help: "Number of active WhatsApp users bridged into Matrix",
+		}),
+		bridgeBlocked: promauto.NewGauge(prometheus.GaugeOpts{
```
Comment (jaller94) [inline metrics.go:113]: Is there not a boolean type? Do we have to use a Gauge?

**Review by jaller94 — CHANGES_REQUESTED** [review]
Looks good from what I know (and just learned) about Go.
I don't think we're going to use the `minDays = 0` right away, but I want to flag that it wouldn't work as I expected it to work.

File: metrics.go (hunk 2 above)
Comment (Half-Shot) [reply metrics.go:113]: Prometheus doesn't have the concept of a boolean. A gauge allows you to go up and down, so is the best fit.

File: main.go (hunk 1 above)
Comment (Half-Shot) [reply main.go:283]: Yeah, I was wondering about this. I think zero is likely a valid value, perhaps we should just always calculate this figure.

Comment (Half-Shot) [reply main.go]: Done

File: config/config.go [hunk 3]
```diff
@@ -60,6 +60,12 @@ type Config struct {
 			Avatar      string `yaml:"avatar"`
 		} `yaml:"bot"`
 
+		Limits struct {
+			MaxPuppetLimit      uint `yaml:"max_puppet_limit"`
+			MinPuppetActiveDays uint `yaml:"min_puppet_activity_days"`
+			BlockOnLimitReached bool `yaml:"block_on_limit_reached"`
```
Comment (Half-Shot) [inline config/config.go:66]: We also need to allow users to drop off after N days. Perhaps `max_puppet_inactivity_days`

**Review by jaywink — CHANGES_REQUESTED** [review]
Seems sensible, though you'll also need to add `PuppetInactivityDays` to the config defaults I would suspect, like the other new items?

File: portal.go [hunk 4]
```diff
@@ -1393,6 +1397,9 @@ func (portal *Portal) HandleTextMessage(source *User, message whatsapp.TextMessa
 	} else {
 		portal.finishHandling(source, message.Info.Source, resp.EventID)
 	}
+	sender := portal.bridge.GetPuppetByJID(message.Info.SenderJid)
+	sender.UpdateActivityTs(message.Info.Timestamp)
```
Comment (Half-Shot) [inline portal.go:1401]: We want to be careful not to store *historical* message times.

File: config/config.go [hunk 5]
```diff
@@ -60,6 +60,13 @@ type Config struct {
 			Avatar      string `yaml:"avatar"`
```
Comment (Half-Shot) [inline config/config.go:60]: TODO: Add a notice to the admin room

**Pull request closed without merging**